src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/CompositeValueProvider.cs

```diff
@@ -52,7 +54,38 @@ public CompositeValueProvider(IList<IValueProvider> valueProviders)
             }
 
             var factories = controllerContext.ValueProviderFactories;
-            var valueProviderFactoryContext = new ValueProviderFactoryContext(controllerContext);
+
+            return await CreateAsync(controllerContext, factories);
+        }
+
+        /// <summary>
+        /// Asynchronously creates a <see cref="CompositeValueProvider"/> using the provided
+        /// <paramref name="actionContext"/>.
+        /// </summary>
+        /// <param name="actionContext">The <see cref="ActionContext"/> associated with the current request.</param>
+        /// <returns>
+        /// A <see cref="Task{TResult}"/> which, when completed, asynchronously returns a
+        /// <see cref="CompositeValueProvider"/>.
+        /// </returns>
+        public static async Task<CompositeValueProvider> CreateAsync(ActionContext actionContext)
+        {
+            if (actionContext == null)
+            {
+                throw new ArgumentNullException(nameof(actionContext));
+            }
+
+            var mvcOptions = actionContext.HttpContext.RequestServices.GetRequiredService<IOptions<MvcOptions>>();
```

Comment:
Any time you use `GetRequiredService` you have to pay @davidfowl a dollar.

If you see yourself using it, then stop and think about it. It's usually wrong.